### PR TITLE
Rewrite of P2P control flow

### PR DESCRIPTION
**Pull request by fjetter**
This builds on https://github.com/dask/distributed/pull/7195 which I was not able to finish because I ran into concurrency issues due to leaky tests. Particularly the global memory_limits on the MultiFile and MultiComm classes caused problems.

I went ahead and rewrote major parts of the concurrency model

- MultiComm and MultiFile are now called `CommShardsBuffer` and `DiskShardsBuffer` respectively
- Both Disk and Comm buffers inherit from a common base class `ShardsBuffer` that implements the concurrency and limiting of both classes
- Both buffers are operating on the event loop such that there is no longer need for threading synchronization primitives.
- To enable this, `Shuffle` needed to be rewritten slightly and uses another offload (specifically in `Shuffle.add_partitions`). We're still doing all the compute on a thread but it's a different thread now. The actual worker thread is mostly idle now.
- No need for any queues any more to control concurrency. Instead, the buffers initialize N coroutines that run until the buffer is closed
- Memory limiting is synchronized with a new `ResourceLimiter` primitive. This is a fairly simple class that allows one to acquire as much of a given resources as one wants but allows us to wait until the acquired resource drops below the specified level. This behavior represents our buffer usage pretty well. Blocking on acquire is not applicable since the memory is already in memory. However, after this acquire we'd like to wait for the memory to calm down before releasing again. This is effectively the same backpressure protocol as on main. This also allows for more flexible control over the buffer size (e.g. comm + disk buffer could have the same limit if we wanted to) and allows for much safer testing by now leaking global state
- Last but not least, concurrency control is mostly implemented using Conditions instead of polling. That may be a subjectively a bit harder to read than the polling that's been there before but allows for significantly faster test runtimes.

I haven't touched serialization, worker dispatching, etc. This is all merely about control flow and responsibility. In follow up PRs I intend to modify actual behavior (e.g. fail on leaving workers)

cc @hendrikmakait @mrocklin

### CODE REVIEW

File: distributed/shuffle/_shuffle_extension.py [hunk 1]
```diff
@@ -37,119 +41,181 @@
...
+                "buckets": len(self._comm_buffer.shards),
+                "written": self._comm_buffer.bytes_written,
                 "read": self.total_recvd,
-                "active": self.multi_comm.queue.qsize(),
-                "diagnostics": self.multi_comm.diagnostics,
-                "memory_limit": self.multi_comm.memory_limit,
+                "active": 0,
+                "diagnostics": self._comm_buffer.diagnostics,
+                "memory_limit": self._comm_buffer.memory_limit,
             },
```
Comment (fjetter) [inline distributed/shuffle/_shuffle_extension.py:198]: - [ ] TODO: correct metrics, otherwise dashboard is broken

**Comment by mrocklin** [comment]
As a heads-up I may not have time to review this during the next couple of days (prepping presentations, travelling, delivering presentations).  The opening comment seems not-scary to me though.  I wouldn't block on my review if you're feeling confident.

File: distributed/shuffle/_shuffle_extension.py [hunk 2]
```diff
@@ -252,30 +328,24 @@ async def shuffle_receive(
...
-            > shuffle.multi_file.memory_limit
-        ):
-            await task  # backpressure
-
-    async def shuffle_inputs_done(self, comm: object, shuffle_id: ShuffleId) -> None:
+        await shuffle.receive(data)
+
+    async def shuffle_inputs_done(self, shuffle_id: ShuffleId) -> None:
         """
         Hander: Inform the extension that all input partitions have been handed off to extensions.
```
Comment (hendrikmakait) [inline distributed/shuffle/_shuffle_extension.py:335]: Driveby: 
```suggestion
        Handler: Inform the extension that all input partitions have been handed off to extensions.
```
(same in L327)

File: distributed/shuffle/_buffer.py [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+            if not self._exception:
+                assert not self.bytes_memory, (type(self), self.bytes_memory)
+
+    async def close(self) -> None:
+        await self.flush()
+        if not self._exception:
+            assert not self.bytes_memory, (type(self), self.bytes_memory)
+        for t in self._tasks:
+            t.cancel()
+        self._closed = True
```
Comment (hendrikmakait) [inline distributed/shuffle/_buffer.py:206]: `flush` also sets `_closed` and `_done`. Should we clean this up so that `flush` is not responsible for setting these, just for checking that the buffer is indeed closed?

File: distributed/shuffle/tests/test_shuffle.py [hunk 4]
```diff
@@ -469,3 +476,208 @@ async def test_clean_after_close(c, s, a, b):
...
+    for ix in range(n_input_partitions):
+        df = pd.DataFrame({"x": range(rows_per_df * ix, rows_per_df * (ix + 1))})
+        df["_partition"] = df.x % npartitions
+        dfs.append(df)
+
+    workers = list("abcdefghijklmn")[:n_workers]
+
+    worker_for_mapping = {}
+
+    for part in range(npartitions):
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:565]: nit for readability:

```suggestion
    for output_partition in range(npartitions):
```
(requires similar adjustments below)

File: distributed/shuffle/tests/test_shuffle.py [hunk 5]
```diff
@@ -469,3 +476,208 @@ async def test_clean_after_close(c, s, a, b):
...
+        total_bytes_recvd = 0
+        total_bytes_recvd_shuffle = 0
+        for s in shuffles:
+            metrics = s.heartbeat()
+            assert metrics["comm"]["total"] == metrics["comm"]["written"]
+            total_bytes_sent += metrics["comm"]["written"]
+            total_bytes_recvd += metrics["disk"]["total"]
+            total_bytes_recvd_shuffle += s.total_recvd
+
+        assert total_bytes_recvd_shuffle == total_bytes_sent
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:605]:
```suggestion
        assert total_bytes_recvd_shuffle == total_bytes_sent == total_bytes_recvd
```

From what I understand, we probably want to test that as well.

File: distributed/shuffle/tests/test_shuffle.py [hunk 6]
```diff
@@ -469,3 +476,208 @@ async def test_clean_after_close(c, s, a, b):
...
+        df_after = pd.concat(all_parts)
+        assert all(_done())
+    finally:
+        await asyncio.gather(*[s.close() for s in shuffles])
+    assert len(df_after) == len(pd.concat(dfs))
+
+
+@gen_test()
+async def test_error_offload(tmpdir, loop_in_thread):
+    # Ensure that all data is
```
Comment (hendrikmakait) [inline distributed/shuffle/tests/test_shuffle.py:628]: Something seems to be missing here

**Review by hendrikmakait — COMMENTED** [review]
Generally, I like this change, the new separation of concerns looks like a great improvement. I have a few questions about minor parts of this PR as well as a number of small nits and suggestions.

File: distributed/shuffle/_buffer.py [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+class ShardsBuffer(Generic[ShardType]):
+
+    shards: defaultdict[str, list[ShardType]]
+    sizes: defaultdict[str, int]
+    _exception: None | Exception
+
+    _queues: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
+
+    @property
+    def _instances(self) -> set:
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:33]: Dead code. Should also be removed

File: distributed/shuffle/_buffer.py [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+        self.bytes_total = 0
+        self.bytes_memory = 0
+        self.bytes_written = 0
+        self.bytes_read = 0
+
+    def heartbeat(self) -> dict[str, Any]:
+        return {
+            "memory": self.bytes_memory,
+            "total": self.bytes_total,
+            "buckets": len(self.shards),
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:66]: shards is a mapping from buckets to shards. buckets are basically the partition IDs

File: distributed/shuffle/_buffer.py [hunk 9]
```diff
@@ -0,0 +1,231 @@
...
+            "written": self.bytes_written,
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:67]: This is all private and I'm inclined to not clean this up for now

File: distributed/shuffle/_buffer.py [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+            self.diagnostics["avg_duration"] = 0.98 * self.diagnostics[
+                "avg_duration"
+            ] + 0.02 * (stop - start)
+        finally:
+            if self.memory_limiter:
+                await self.memory_limiter.release(size)
+            self.bytes_memory -= size
+
+    async def _process(self, id: str, shards: list[ShardType]) -> None:
+        raise NotImplementedError()
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:97]: What would be the benefit of this?

File: distributed/shuffle/_buffer.py [hunk 11]
```diff
@@ -0,0 +1,231 @@
...
+
+    def read(self, id: str) -> ShardType:
+        raise NotImplementedError()
+
+    @property
+    def empty(self) -> bool:
+        return not self.shards
+
+    async def worker(self) -> None:
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:106]: yes, already staged

File: distributed/shuffle/_shuffle_extension.py [hunk 12]
```diff
@@ -37,119 +41,167 @@
...
-        return await asyncio.get_event_loop().run_in_executor(
-            self.executor,
-            func,
-            *args,
+    async def barrier(self) -> None:
+        # FIXME: This should restrict communication to only workers
+        # participating in this specific shuffle. This will not only reduce the
+        # number of workers we need to contact but will also simplify error
+        # handling, e.g. if a non-participating worker is not reachable in time
+        # TODO: Consider broadcast pinging once when the shuffle starts to warm
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_shuffle_extension.py:153]: yes

File: distributed/shuffle/tests/test_shuffle.py (hunk 5 above)
Comment (fjetter) [reply distributed/shuffle/tests/test_shuffle.py:605]: these are not identical. there appears to be a small drift between sizes when comparing pyarrow buffers and bytes directly. Therefore, what the comms measure is slightly different from what is actually received. This is very ugly but I'm not willing to fix this right now. Refactoring the serialization parts is out of scope

File: distributed/shuffle/_buffer.py (hunk 3 above)
Comment (fjetter) [reply distributed/shuffle/_buffer.py:206]: In a perfect world, flush would not necessarily close the buffer at all. however, untangling this proved to be a bit more difficult than I thought it would be. I'll keep this as is, for now. This PR is already big enough as it is.

File: distributed/shuffle/_buffer.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+        def _continue() -> bool:
+            return bool(self.shards or self._done)
+
+        while True:
+            async with self._shards_available:
+                await self._shards_available.wait_for(_continue)
+                if self._done and not self.shards:
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:113]: if an exception is raised, done is already set.

File: distributed/shuffle/_buffer.py [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+        concurrency_limit: int = 2,
+        max_message_size: int = -1,
+    ) -> None:
+        self._closed = False
+        self.shards = defaultdict(list)
+        self.sizes = defaultdict(int)
+        self._exception = None
+        self.concurrency_limit = concurrency_limit
+        self._done = False
+        self.memory_limiter = memory_limiter
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:48]: I think mypy can figure this out on it's own

File: distributed/shuffle/tests/test_shuffle.py (hunk 5 above)
Comment (hendrikmakait) [reply distributed/shuffle/tests/test_shuffle.py:605]: Let's highlight this in the test then or drop `total_bytes_recvd` altogether. Right now it looks like we just forgot to do something useful with `total_bytes_recvd`.

File: distributed/shuffle/_buffer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/shuffle/_buffer.py:66]: Let's maybe rename `self.shards` to `self.bucketed_shards` or `self.partitioned_shards` to highlight that.

File: distributed/shuffle/_buffer.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/shuffle/_buffer.py:97]: Marking `ShardsBuffer` as an `abc.ABC` as well as decorating `_process` with `@abc.abstractmethod` feels mildly cleaner since `ShardsBuffer` should not be initialized and any concrete subclass needs to implement `_process`[1]. This would allow `mypy` to catch errors such as trying to instantiate `ShardsBuffer` or a subclass that did not implement `_process`. Similarly, Python would throw somewhat more informative errors at runtime.

I guess this is mainly a question about conventions: If a class is an abstract class, should it always be marked as such or do we not care about that. If it is the latter, why do we not care? (We don't see any value in this case is a perfectly valid answer).

Feel free to shut this discussion down, the case at hand revolves around an abstract class that will have two concrete subclasses for the foreseeable future and is in a private module, so I'm rather interested in whether we have some conventions/guidelines around marking ABCs, not splitting hairs about this specific case.

[1] `read` not so much, which leads me off-trail to asking whether `read` belongs here in the class hierarchy, but that's a fringe problem that requires more changes and I don't want to deal with this right now.

File: distributed/shuffle/tests/test_buffer.py [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+        buf.allow_process.clear()
+
+        many_small = [asyncio.create_task(buf.put(small_payload)) for _ in range(11)]
+        assert buf.memory_limiter
+        while buf.memory_limiter.available():
+            await asyncio.sleep(0.1)
+
+        new_put = asyncio.create_task(buf.put(small_payload))
+        with pytest.raises(asyncio.TimeoutError):
+            await asyncio.wait_for(asyncio.shield(new_put), 0.1)
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/tests/test_buffer.py:74]: yes

File: distributed/shuffle/tests/test_limiter.py [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+    await res.release(1)
+    assert not res.available()
+
+    with pytest.raises(asyncio.TimeoutError):
+        await asyncio.wait_for(res.wait_for_available(), 0.1)
+
+    await res.release(1)
+    assert res.available() == 1
+    await res.wait_for_available()
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/tests/test_limiter.py:34]: done (pending push)

File: distributed/shuffle/tests/test_buffer.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+
+    async with BufferTest(
+        memory_limiter=limiter,
+        concurrency_limit=2,
+    ) as buf:
+        many_small = [asyncio.create_task(buf.put(small_payload)) for _ in range(9)]
+        buf.allow_process.set()
+        many_small = asyncio.gather(*many_small)
+        # Puts that do not breach the limit do not block
+        await asyncio.wait_for(many_small, 0.1)
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/tests/test_buffer.py:64]: I added a bit of logic to the limiter to track blocked time. this way we can clearly assert if there was any block at all no matter how small and we can remove the wait_for here

File: distributed/shuffle/_buffer.py [hunk 18]
```diff
@@ -0,0 +1,231 @@
...
+                                del self.shards[part_id]
+                                assert not self.sizes[part_id]
+                                del self.sizes[part_id]
+                else:
+                    shards = self.shards.pop(part_id)
+                    size = self.sizes.pop(part_id)
+                self._shards_available.notify_all()
+            await self.process(part_id, shards, size)
+
+    async def put(self, data: dict[str, list[ShardType]]) -> None:
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:139]: changed to write

File: distributed/shuffle/_buffer.py [hunk 19]
```diff
@@ -0,0 +1,245 @@
...
+
+    Buckets typically correspond to output partitions.
+
+    If exceptions occur during writing, the buffer is automatically closed. Subsequent attempts to write will raise the same exception.
+    Flushing will not raise an exception. To ensure that the buffer finished successfully, please call `ShardsBuffer.raise_on_exception`
+    """
+
+    shards: defaultdict[str, list[ShardType]]
+    sizes: defaultdict[str, int]
+    _exception: None | Exception
```
Comment (hendrikmakait) [inline distributed/shuffle/_buffer.py:44]: nit: Complete annotations for attributes
```suggestion
    shards: defaultdict[str, list[ShardType]]
    sizes: defaultdict[str, int]
    concurrency_limit: int
    memory_limiter: ResourceLimiter | None
    diagnostics: dict[str, float]
    max_message_size: int

    bytes_total: int
    bytes_memory: int
    bytes_written: int
    bytes_read: int

    _closed: bool
    _done: bool
    _exception: None | Exception
    _tasks: list[asyncio.Task]
    _shards_available: asyncio.Condition
    _flush_lock: asyncio.Lock
```

File: distributed/shuffle/_buffer.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/shuffle/_buffer.py:58]: True, it's just nice to have a standardized spot to look for annotations and not rely on `mypy` as mentioned in https://github.com/dask/community/issues/255. I've added a suggestion that extends the annotations, feel free to ignore.

File: distributed/shuffle/_buffer.py [hunk 20]
```diff
@@ -0,0 +1,231 @@
...
+            asyncio.create_task(self.worker()) for _ in range(concurrency_limit)
+        ]
+        self._shards_available = asyncio.Condition()
+        self._flush_lock = asyncio.Lock()
+        self.max_message_size = max_message_size
+
+        self.bytes_total = 0
+        self.bytes_memory = 0
+        self.bytes_written = 0
+        self.bytes_read = 0
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:71]: I'm not sure, yet, what will happen to this. the future of diagnostics is also unclear. moving everything around and organizing feels a bit premature and I'd rather focus on more impactful things now

File: distributed/shuffle/_buffer.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_buffer.py:77]: IMO this is OK. We're not naming every single mapping as `key_value`. Context is typically sufficient to infer this. In this case I think the shards mapping is fine and a more verbose attribute name make the code harder to read

File: distributed/shuffle/_limiter.py [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+import asyncio
+
+
+class ResourceLimiter:
+    def __init__(self, maxvalue: int) -> None:
+        self._maxvalue = maxvalue
+        self._acquired = 0
+        self._condition = asyncio.Condition()
+        self._waiters = 0
+        self._backlog_size = 0
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/shuffle/_limiter.py:30]: backlog size stems from an earlier implementation. waiters is not used, yet, but I'll keep it anyhow. this is useful for dashboards, etc. (I have something in the pipeline but not ready for a PR,y et)

**Review by hendrikmakait — APPROVED** [review]
Thanks for adding the additional documentation, @fjetter. This looks good to me, my remaining questions/nits can be safely ignored.
